cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll

```diff
@@ -381,3 +383,135 @@ Function resolveCall(Call call) {
     result = Dispatch::viableCallable(callInstruction)
   )
 }
+
+/**
+ * Provides definitions for augmenting source/sink pairs with data-flow paths
+ * between them. From a `@kind path-problem` query, import this module in the
+ * global scope, extend `TaintTrackingConfiguration`, and use `taintedWithPath`
+ * in place of `tainted`.
+ *
+ * Importing this module will also import the query predicates that contain the
+ * taint paths.
+ */
+module TaintedWithPath {
+  /**
+   * A taint-tracking configuration that matches sources and sinks in the same
+   * way as the `tainted` predicate.
+   */
+  class TaintTrackingConfiguration extends int {
+    TaintTrackingConfiguration() { this = 1 }
+
+    /** Override this to specify which elements are sinks in this configuration. */
+    abstract predicate isSink(Element e);
+  }
+
+  private class AdjustedConfiguration extends DataFlow3::Configuration {
+    AdjustedConfiguration() { this = "AdjustedConfiguration" }
+
+    override predicate isSource(DataFlow::Node source) { source = getNodeForSource(_) }
+
+    override predicate isSink(DataFlow::Node sink) {
+      exists(TaintTrackingConfiguration cfg | cfg.isSink(adjustedSink(sink)))
+    }
+
+    override predicate isAdditionalFlowStep(DataFlow::Node n1, DataFlow::Node n2) {
+      instructionTaintStep(n1.asInstruction(), n2.asInstruction())
+    }
+
+    override predicate isBarrier(DataFlow::Node node) { nodeIsBarrier(node) }
+
+    override predicate isBarrierIn(DataFlow::Node node) { nodeIsBarrierIn(node) }
+  }
+
+  /*
+   * A sink `Element` may map to multiple `DataFlowX::PathNode`s via (the
+   * inverse of) `adjustedSink`. For example, an `Expr` maps to all its
+   * conversions, and a `Variable` maps to all loads and stores from it. Because
+   * the path node is part of the tuple that constitutes the alert, this leads
+   * to duplicate alerts.
+   *
+   * To avoid showing duplicates, we edit the graph to replace the final node
+   * coming from the data-flow library with a node that matches exactly the
+   * `Element` sink that's requested.
+   *
+   * The same should ideally be done with the source, but we haven't seen a
+   * need for it yet.
```

Comment:
It's mapped to the (unique) unconverted expression, so each `Expr` source is mapped to just one `DataFlow::Node` source. That means there's no ambiguity, unlike `adjustedSink`, where each `Element` sink can map to many `DataFlow::Node` sinks.

There is a bit of ambiguity in `getNodeForSource` because an `Expr` source can become both an `ExprNode` and a `DefinitionByReferenceNode`. But that ambiguity if part of what I'm trying to debug at the moment because it causes other problems than duplicate alerts. A partial fix was in #3137.

Reply:
Each `DataFlow::Node` maps to a unique unconverted expression, but an `Expr` source can map to multiple `DataFlow::Node` sources if it gets converted. It's possible that the existing tests don't exercise that, but I expect it to be a problem on real code.

Reply:
I think you're right. I've now seen a few tests that are affected by this, although in almost all cases it's mitigated by our rule to suppress flow into a source:

https://github.com/Semmle/ql/blob/604731ba6b06b43e31c9891d1471bcd8b9dce936/cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll#L169-L172

These rules together mean that the first step in a path explanation is often a `Conversion`, which looks a bit confusing.

Reply:
Done.